components/camera/rtsp/rtsp.go

```diff
@@ -0,0 +1,240 @@
+}
+
+// Validate checks to see if the attributes of the model are valid.
+func (at *Attrs) Validate() error {
+	if prefix := strings.HasPrefix(at.Address, "rtsp://"); !prefix {
+		return errors.New(`rtsp_address must begin with "rtsp://"`)
+	}
+	if err := at.IntrinsicParams.CheckValid(); err != nil {
+		return err
+	}
+	if err := at.DistortionParams.CheckValid(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// rtspCamera contains the rtsp client, and the reader function that fulfills the camera interface.
+type rtspCamera struct {
+	gostream.VideoReader
+	u                       *url.URL
+	client                  *gortsplib.Client
+	cancelCtx               context.Context
+	cancelFunc              context.CancelFunc
+	activeBackgroundWorkers sync.WaitGroup
+	gotFirstFrameOnce       bool
+	latestFrame             atomic.Value
+	logger                  golog.Logger
+}
+
+// Close closes the camera. It always returns nil, but because of Close() interface, it needs to return an error.
+func (rc *rtspCamera) Close(ctx context.Context) error {
+	rc.cancelFunc()
+	rc.activeBackgroundWorkers.Wait()
+	if err := rc.client.Close(); err != nil && !errors.Is(err, clientTerminated) {
+		rc.logger.Infow("error while closing rtsp client:", "error", err)
+	}
+	return nil
+}
+
+// clientReconnectBackgroundWorker checks every 5 sec to see if the client is connected to the server, and reconnects if not.
+func (rc *rtspCamera) clientReconnectBackgroundWorker() {
+	rc.activeBackgroundWorkers.Add(1)
+	ticker := time.NewTicker(5 * time.Second)
+	goutils.ManagedGo(func() {
+		for {
+			if ok := goutils.SelectContextOrWaitChan(rc.cancelCtx, ticker.C); ok {
+				// use an OPTIONS request to see if the server is still responding to requests
+				res, err := rc.client.Options(rc.u)
+				if err != nil && (errors.Is(err, clientTerminated) ||
+					errors.Is(err, io.EOF) ||
+					errors.Is(err, syscall.EPIPE) ||
+					errors.Is(err, syscall.ECONNREFUSED)) {
+					rc.logger.Warnw("The rtsp client encountered an error, trying to reconnect", "url", rc.u, "error", err)
+					if err = rc.reconnectClient(); err != nil {
+						rc.logger.Warnw("cannot reconnect to rtsp server", "error", err)
+					} else {
+						rc.logger.Infow("reconnected to rtsp server", "url", rc.u)
+					}
+				} else if res != nil && res.StatusCode != base.StatusOK {
+					rc.logger.Warnw("The rtsp server responded with non-OK status", "url", rc.u, "status code", res.StatusCode)
+					if err = rc.reconnectClient(); err != nil {
+						rc.logger.Warnw("cannot reconnect to rtsp server", "error", err)
+					} else {
+						rc.logger.Infow("reconnected to rtsp server", "url", rc.u)
+					}
+				}
+			} else {
+				return
+			}
+		}
+	}, rc.activeBackgroundWorkers.Done)
+}
+
+// reconnectClient reconnects the RTSP client to the streaming server by closing the old one and starting a new one.
+func (rc *rtspCamera) reconnectClient() (err error) {
+	if rc.client != nil {
```

Comment:
nit: this is a SEGV if `rc == nil`, though looking at the code, I don't think that'll ever be the case. Feel free to ignore.

Reply:
done